src/mpid/ch4/include/mpidpre.h

```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
     uint16_t type;
 };
 
-struct MPIDI_CH4U_win_queue {
+typedef struct MPIDI_CH4U_win_lock_recvd {
     struct MPIDI_CH4U_win_lock *head;
     struct MPIDI_CH4U_win_lock *tail;
-};
+    int type;                   /* current lock's type */
+    unsigned count;             /* count of granted locks (not received) */
+} MPIDI_CH4U_win_lock_recvd_t;
 
-typedef struct MPIDI_CH4U_win_lock_info {
-    unsigned peer;
-    int lock_type;
-    struct MPIR_Win *win;
-    volatile unsigned done;
-} MPIDI_CH4U_win_lock_info;
+typedef struct MPIDI_CH4U_win_target_sync_lock {
+    volatile unsigned locked;   /* locked == 0 or 1 */
```

Comment:
I don't get the "inlining" explanation.
1. Considering single threaded execution, compiler should be able to recognize dependencies correctly even if functions are inlined. But apparently you are talking about multi-threaded case.
2. Then in multi-threaded execution, every time a thread accesses (either read or write) the variable, it must hold a lock. Whenever read happens to the variable, the thread needs to acquire the lock beforehand. `pthread_mutex_lock` is not inlined, and even if it was inlined, it is supposed to contain both compiler/memory barrier to prevent ordering change across this lock operation. And of course once the thread takes a lock, you don't have to worry about other thread modifying the variable.